library/src/scala/util/boundary.scala

```diff
@@ -0,0 +1,41 @@
+package scala.util
+
+/** A boundary that can be exited by `break` calls.
+ *  `boundary` and `break` represent a unified and superior alternative for the
+ *  `scala.util.control.NonLocalReturns` and `scala.util.control.Breaks` APIs.
+ *  The main differences are:
+ *
+ *    - Unified names: `boundary` to establish a scope, `break` to leave it.
+ *      `break` can optionally return a value.
+ *    - Integration with exceptions. `break`s are logically non-fatal exceptions.
+ *      The `Break` exception class extends `RuntimeException` and is optimized so
+ *      that stack trace generation is suppressed.
+ *    - Better performance: breaks to enclosing scopes in the same method can
+ *      be rwritten to jumps.
+ */
+object boundary:
+
+  /** User code should call `break.apply` instead of throwing this exception
+   *  directly.
+   */
+  final class Break[T] private[boundary](val label: Label[T], val value: T)
```

Comment:
This class should probably be in `object break`. This would make the `Label.break` method redundant.

```scala
object break:
  final class Break[T] private[break](val label: Label[T], val value: T) ...
  def apply[T](value: T)(using label: boundary.Label[T]): Nothing =
    throw Break(label, value)
  def apply()(using label: boundary.Label[Unit]): Nothing =
    apply(())
```

Reply:
In fact, I'd say the other way around: `break` should be in `boundary` (as two overloaded `def break` methods). The recommended pattern should be
```scala
boundary {
  ...
  boundary.break(v)
  ...
}
```
Given that `boundary` should be used quite sparingly, mostly as a building block for other things, there is no reason to pollute the global `scala.util.*` scope with a generic name `break`.

This also allows things to be `private[boundary]`, and we're sure to keep things as localized as possible.

Reply:
I agree it's best to combine all essential pieces in the `boundary` object. I'd tend to leave `break` as an inline method in `scala.util`, so that it's easier to use. After all, we do propose this as a replacement for nonlocal returns, so we should be able to use it without major import acrobatics.

Reply:
I believe that writing `boundary.break(v) ` is better because it clearly shows the link between this method and the boundary. An extra import is only useful if the user has to write many breaks. 

By adding `break` as an object we are also adding more conceptual indirections as the methods they would use are the `apply` methods. If we would define those as methods in `boundary`, users would be able to see the complete documentation in the `boundary`.

Either way, `Break` and `break` should be defined together. I also don't see to point in adding a `private[util] def break` when `break.apply` could just throw the `Break` directly.

Reply:
I removed the `break` in `scala.util`. Everything is now in `boundary`.